trio/testing/_checkpoints.py

```diff
@@ -13,26 +13,27 @@ def _assert_yields_or_not(expected):
     orig_schedule = task._schedule_points
     try:
         yield
-    finally:
         if (
             expected and (
                 task._cancel_points == orig_cancel
                 or task._schedule_points == orig_schedule
             )
         ):
             raise AssertionError("assert_checkpoints block did not yield!")
-        elif (
+    finally:
+        if (
             not expected and (
                 task._cancel_points != orig_cancel
                 or task._schedule_points != orig_schedule
             )
         ):
-            raise AssertionError("assert_no_yields block yielded!")
+            raise AssertionError("assert_no_checkpoints block yielded!")
 
 
 def assert_checkpoints():
     """Use as a context manager to check that the code inside the ``with``
-    block executes at least one :ref:`checkpoint <checkpoints>`.
+    block either exits with an exception or executes at least one
+    :ref:`checkpoint <checkpoints>`.
```

Comment:
Technically this is a user-visible change in semantics too. I doubt anyone much will care, but we should probably mention it in the release notes anyway.